File: eskybe/src/main/java/com/esky/service/CourseService.java
package com.esky.service;

import com.esky.model.entities.Course;
import com.esky.model.pojo.CourseRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface CourseService {

    //Save Course:
    public Course saveCourse(CourseRequest course);

    //Update Course:
    public Course updateCourse(CourseRequest course);

    //Delete Course:
    public void deleteCourse(CourseRequest course);

    //Delete Course by ID:
    public void deleteCourseById(Long id);

    //Get Course by ID:
    public Course getCourseById(Long id);

    //Get all Courses by Filter:
    public Page<Course> findAllCourseByFilter(PageRequest pageRequest, String filterValue);

    //Get all Courses by Creator:
    public Page<Course> findAllCourseByCreator(PageRequest pageRequest, String filterValue, Long idCreator);

    //Get all Courses by Student:
    public Page<Course> findByStudentAndFilter(PageRequest pageRequest, String filterValue, Long idCreator);

    //Get my Courses by Creator:
    public List<CourseRequest> findMyCourseByCreator(Long idCreator);

    //Find all Courses:
    public Page<Course> findAllCourse(PageRequest pageRequest);

    //Get Courses Count
    public long getCoursesCount();

}

File: webui/src/main/webui/src/app/shared/services/http/course.service.ts
import {Injectable} from '@angular/core';
import {HttpClient, HttpParams} from '@angular/common/http';
import {HttpHeaders} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Course} from "../../entities/course";
import {api, environment} from "../../../../environments/environment";
import {User} from "../../entities/user";
import {Pageable} from "../../entities/pageable";

const httpOptions = {
    headers: new HttpHeaders({
        'Content-Type': 'application/json',
        'Authorization': 'my-auth-token'
    })
};

@Injectable()
export class CourseService {

    constructor(private http: HttpClient) {
    }

    getCountCourses() {
        return this.http.get<number>(environment.endpoint + api.course + api.count);
    }

    getAllCourses(pageable: Pageable) {
        const params = new HttpParams()
            .set('page', pageable.page)
            .set('size', pageable.size)
            .set('direction', pageable.direction)
            .set('sort', pageable.sort);
        return this.http.get<Course[]>(environment.endpoint + api.course + api.all, {params});
    }

    getCoursesByFilter(pageable: Pageable, filterValue: string): any {
        const params = new HttpParams()
            .set('page', pageable.page)
            .set('size', pageable.size)
            .set('direction', pageable.direction)
            .set('sort', pageable.sort)
            .set('filterValue', filterValue);
        return this.http.get<Course[]>(environment.endpoint + api.course + api.all + api.filter, {params});
    }

    getCoursesByCreator(pageable: Pageable, filterValue: string, idCreator: number): any {
        const params = new HttpParams()
            .set('page', pageable.page)
            .set('size', pageable.size)
            .set('direction', pageable.direction)
            .set('sort', pageable.sort)
            .set('filterValue', filterValue);
        return this.http.get<Course[]>(environment.endpoint + api.course + api.all + api.creator + idCreator, {params});
    }

    getMyCoursesByCreator(idCreator: number): any {
        return this.http.get<Course[]>(environment.endpoint + api.course + api.mine + api.creator + idCreator);
    }

    getCoursesByStudent(pageable: Pageable, filterValue: string, idStudent: number): any {
        const params = new HttpParams()
            .set('page', pageable.page)
            .set('size', pageable.size)
            .set('direction', pageable.direction)
            .set('sort', pageable.sort)
            .set('filterValue', filterValue);
        return this.http.get<Course[]>(environment.endpoint + api.course + api.all + api.student + idStudent, {params});
    }

    getCourseByCreator(creator: User) {
        return this.http.get<Course>(environment.endpoint + api.course + api.get + creator);
    }

    getCourseById(id: number) {
        return this.http.get<Course>(environment.endpoint + api.course + api.get + id);
    }

    addCourse(course: Course): Observable<Course> {
        return this.http.post<Course>(environment.endpoint + api.course + api.save, course, httpOptions);
    }

    updateCourse(course: Course): Observable<Course> {
        console.log(JSON.stringify(course.id));
        return this.http.put<Course>(environment.endpoint + api.course + api.update + course.id, course, httpOptions);
    }

    deleteCourse(id: number) {
        return this.http.delete<Course>(environment.endpoint + api.course + api.delete + id);
    }

    getCourseCount(type: any, id: any) {
        return this.http.get<number>(environment.endpoint + api.course + '/' + type + api.count + id);
    }
}